### PR TITLE
Add Mix recompilation from IEx

### PR DESCRIPTION
**Pull request by josevalim**
Because this is asked in the mailing list every other week,
I have decided to take a stab at it.

I went with the more conservative approach of always restarting
the application so users don't see errors from their processes.

There are a bunch of disclaimers to consider which have been
documented. For this reason, the helper has been marked as
experimental so we see how often folks will run into corner cases.

@ericmj, @fishcakez and @alco: careful review and feedback is
appreciated. I was also torn between calling it restart and recompile
but went with the latter as the need for restarting is a side-effect
of compilation.

Thoughts?

### CODE REVIEW

**Comment by fishcakez** [comment]
If we do this we need to close the umbrella apps in the right order. So if MyApp1 depends on MyApp2 (has :my_app2 in applications) then we stop MyApp1 before MyApp2.

File: lib/iex/lib/iex/helpers.ex [hunk 1]
```diff
@@ -50,9 +51,68 @@ defmodule IEx.Helpers do
   To learn more about IEx as a whole, just type `h(IEx)`.
   """
 
+  require Logger
```
Comment (lexmag) [inline lib/iex/lib/iex/helpers.ex]: Needless `require`?

**Comment by josevalim** [comment]
@fishcakez good call. we need to reverse the list (which is properly ordered according to deps).

File: lib/iex/lib/iex/helpers.ex (hunk 1 above)
Comment (josevalim) [inline lib/iex/lib/iex/helpers.ex]: Good catch, fixing it.

**Comment by josevalim** [comment]
Btw, this command will print that the applications are being stopped.

**Pull request merged**